### ISSUE TITLE
Model 12 diagnostic ROM

### ISSUE THREAD

**Issue by joe-novosel** [misterblack1/trs80-diagnosticrom#14]
The diagnostic ROM doesn't work so far on my Model 12.  The system has slimline drives that don't have a head load.  When I start the system, the drive spins and garbage displays on the screen.  

I bought the system in a broken state and got it working -- at least it gives the "boot error" message.  I'm waiting on the parts to connect a floppy emulator to test booting.  I don't have any 8 inch floppy disks yet either.  

The Model 12 and 6000 are the same Z80 board and there is a beeper.  I'm willing to help test for this platform.

**Comment by ki3v** [comment]
Hm.  I did have another person test on a model 6000, and it got through the memory testing OK.

I feel like it shouldn't matter about the head load solenoid, as the FDC should accept the command and the drive should ignore it... you can use non-slimline drive chassises on the model 12, no?

When you say it shows garbage, do you mean garbage, or the character test pattern?  That is, all characters shown repeatedly in increasing order like you can see in the animation on the main page of the repository?  If there is a FDC issue, it can happen that the ROM freezes after showing the character test pattern.  But, if it's true garbage it sounds like the ROM isn't even running.

**Comment by joe-novosel** [comment]
It’s doing something.  Here’s output. It seems to initialize video and drive starts spinning.  


The stock boot rom does this


No ROM at all does nothing. No video and no drive on. 

I’ll setup the logic analyzer next to see where it’s hanging. 

This system was dead when I got it so no telling what’s wrong. 

Thanks
Joe Novosel — N4HC

**Comment by ki3v** [comment]
I don’t see the images, but that could be because I’m on mobile at the moment… I’ll look again as soon as I can from a desktop machine. 

I forgot to ask, what boot error message exactly do you get when trying to boot with the stock ROM? IIRC when you leave the drive empty it should say “insert diskette”, not give a boot error. I can look in the stock boot ROM disassembly for your error message.

**Comment by joe-novosel** [comment]
Boot error MF.  Which I understand to be a memory error.  It’s just not clear which bank.

**Comment by ki3v** [comment]
Yes, you're right, MF means memory error from the built-in ROM.  Its RAM test is rudimentary, but it's obviously failing.

If you watch Adrian's videos, you'll see that he had some problems with one of the chips in the address decoding logic, which caused the FDC chip to be selected in certain situations when it should not have been.  This made the floppy drive become selected and load the heads when the ROM was only changing RAM banks.  I wonder if it's possible that you've got something similar going on, though probably not with the FDC, and possible with memory addresses rather than I/O addresses.  

If that's happening, it could be that certain memory ranges work but others do not, and the ranges might not be contiguous.  If a certain address line is getting decoded wrong in part of the system, for instance, you might have every other block of 256 bytes corrupted on read or write, regardless of whether the correct address is RAM or ROM.  So part of the ROM may work fine, allowing the stock ROM to work as far as it does, but our diag ROM may fail because it runs code at addresses that trigger bus contention (or something similar).

A bad RAM chip alone should not cause the problems you're describing.  The diag ROM can run with all RAM chips removed from the machine.  It can run without VRAM too, though getting the test results from the floppy disk light is pretty tedious.  It does, however, require all of the address lines and chip select logic to be working.

Sadly, I'm still not able to see the images you included, and I'm not sure why.  It looks like you sent your reply from an iPhone.  Can you please log into the GitHub issue page and drag the images in directly as attachments?

I could probably make you a quick custom ROM that does not attempt to access the FDC, to see if an FDC problem has anything to do with it.

Rather than a logic analyzer, I'd suggest you try an analog oscilloscope and probe the address lines while the machine is running... although this can be hard on these Big Tandy machines if you don't have the means to run the board isolated on the benchtop like in Adrian's videos.  But if you can view the address lines on the analog scope, bus contention can be easy to see.

**Comment by joe-novosel** [comment]
David,

I was able to build a copy of the ROM that bypasses the video ram fail that blinks the drive and get to the main test screen.  All of the RAM is “absent.”

I’m thinking the video ram is good – see the test outputs.  For some reason the test is failing.  

(before this I commented out some lines and got the drive to blink – 8 seconds one second on, repeat)

 

In addition while checking and all the (non-PAL) socketed chips (and de-oxit) I found a couple chips with bent pins – floppy controller and Z80-PIO.  So someone before me had tried to fix this and made a mess.  

 

I probed signals with a scope and found that RAS/CAS were missing and traced that to the memory timing circuit where I found a couple of bad 74S112 chips.  (pulled and tested)

 

So far the test has been a huge help once I was able to hack it to jump past the failures.  

 

Thanks,

Joe Novosel

**Comment by joe-novosel** [comment]
![image0](https://user-images.githubusercontent.com/54041911/200445337-50c8c555-2164-4551-9f60-a9b6f9d49e35.jpeg)
![image1](https://user-images.githubusercontent.com/54041911/200445348-d731d45d-f29f-44eb-b0a2-bafc8da77c53.jpeg)

**Comment by ki3v** [comment]
The report of "absent" in reality means that every bit is testing bad.  

There's a shortcut test at the beginning of the RAM test that checks for absent RAM by writing [edit: had left out the word "writing"] all ones to one location then all zeros to the next location, and checking the first location to see if we read back what we wrote to the second one.  If the bus is floating (no RAM chips selected), we'll usually read back whatever we last wrote to the bus, regardless of what address is selected in the missing bank.  This test is very fast and is used to accelerate through genuinely missing banks.

That said, the test will report "absent" any time all 8 bits fail the RAM test.  It just comes back faster when that shortcut test fails.

I continue to suspect either that you have an address line problem like I mentioned before, or maybe a bus transceiver is out resulting in bad data for all RAM locations on your Z80 board.  

It's also interesting that you seem to have some "snow" on the page where the RAM test is running.  Even up on the top where the banner is printed.  It seems to result in a row of pixels being set (or in the case of reversed video, unset).  That is, inside a character cell, rows of the character seem to be randomly coming back as all ones.  What's interesting about that is that the opposite is not true... there don't seem to be characters where there are rows of pixels missing.  Also, the fact that it seems to happen always on character cell boundaries suggests that something is disturbing the VDC chip during the RAM test.

On the second screen you show, the whole display is clean.  So the character ROM itself seems to be working K (and in this machine the chargen ROM is not in the Z80-addressable memory space anyway).  And it's not VRAM corruption or the whole character would be turned to garbage and stay that way.

Troubling about the bent pins etc.  Yes sounds like a gorilla was in the machine.  Can't get too upset, we were all gorillas at this at one point (I probably still qualify as a gorilla).

Missing RAS/CAS certainly is a problem.  You've sorted that out at this point?  Seems a good chance there may still be some bad pins, or possibly bad sockets... remember that sockets can be damaged by abuse, so a previous attempt at poking the machine to life may have damaged a socket and that could be frustrating to find, especially if you're not checking for it.

Good luck, you seem to be making progress!  Not sure if I'm helping at all.  As disclaimed elsewhere, I'm definitely more of a software guy than a hardware guy.

Also: your CRT is nice and sharp!  I sure hope you get that machine running.

**Comment by joe-novosel** [comment]
I pulled all the transceivers and other TTL parts  (that were in sockets at least) and tested them in my TL866II -  all passed.  

I’m ordering some replacement 74S112 chips.  These are part of the RAS/CAS generation.  There’s still no telling what else may be bad once I get the memory fixed.  

 

Yes, I’m getting “snow” when the test is running.  It looks like updates are happening too fast?  The pic could be a camera artefact.  A longer shutter speed would probably help but it’s an iPhone so… 

 

Thanks,

Joe Novosel

**Comment by ki3v** [comment]
"It looks like updates are happening too fast?  The pic could be a camera artefact.  A longer shutter speed would probably help but it’s an iPhone so…"

I don't think it's an artifact from the iPhone.  Generally that would be a whole top/bottom half of the screen dimmer, or a big swath through the middle.

Some of the flickering on the "absent" may be the word "Test" in reverse coming and going before the lines are scanned, so not actually snow.  To be fair, dim lines would be harder to see because of phosphor persistence, so I might be wrong about there being no lines with missing pixels.

But broadly speaking, I don't think the Model II/12/16/6000 are generally susceptible to snow when everything is working properly, so I *think* it's related to something wrong on your machine.  Certainly could be wrong though, so I'd call that an ambiguous hint rather than a smoking gun at this point.

**Comment by misterblack1** [comment]
Chiming in -- there really shouldn't be snow during the regular RAM tests (on the screen) -- the test only updates VRAM while changing text on screen. I would assume your video circuitry has an issue too, which is why that test is failing. You must have bypassed it by just making it think the VRAM is good? 

Perhaps you have a marginal drivers on a bus transceiver, causing both of those systems to "kind of" stomp on each other? It sounds like it's going to be a tricky problem to solve. 

Side note, the TL866 is a terrible chip tester -- it's a CMOS device, so if you have chips with marginal drivers, it doesn't always detect that. I've had chips that were causing issues in machines come back as "good" on the TL866 so I don't generally use it for testing anymore. Usually a "bad" on the TL866 is probably really bad, but it can give you false "goods."

**Comment by misterblack1** [comment]
OK David tells me the new version of the ROM zips through nearly instantly when banks are missing..... so that would explain the snow actually. The old test (that I used) took about 0.5s per bank (even when absent.) 

I still say something is wrong with the VRAM which is why it's hanging there... but clearly it works well enough to show good data so you can probably ignore that for now. Parts of VRAM are not visible on the 80x25 screen -- about 48 bytes of VRAM are not used IIRC .... 80x25 only takes up 2000 characters and I think there are 2048 bytes of VRAM. Could be the bad location is in the non visible part.

**Comment by joe-novosel** [comment]
The TL866 isn't much of an EPROM programmer either...  I just ordered an RCT Pro with a 2716 programming adapter.  
I'm not sure what happens to a 2716 ram when the +12 goes away (fuse blown for motherboard +12V) but I suspect it had some undesirable results.  When I powered it up after replacing fuse the 4116 RAM got too hot to touch.  I'm suspecting any chips that touch the DRAM which includes the bus transceivers.  I'm replacing all with 4164 parts.  The Model 12 has jumpers to allow that option.  

I tested the 2114 chips with an Arduino and they "pass" although it's not an exhaustive test.

**Comment by joe-novosel** [comment]
![IMG_0646](https://user-images.githubusercontent.com/54041911/201528743-171387f8-6e14-4869-a26c-db35e45cb961.jpg)
I got it working.  The video card was causing the problem.  On a whim I tested with the card out, RAS and CAS were generated.  I changed jumpers on the video card (E2-E3, E4-E5) to only allow CPU access during vertical and horizontal retrace and this resolved the issue.  This isn't the default configuration.  I may still dig around the video card to see if there are bad chips there but as long as the system works I probably won't.  

I still had bad chips (RAM, CPU at minimum) and some with pins bent.  At this point the diagnostic passes.  Time to move on to reassembly.  Having this code with source that I could tweak was very valuable.  Thanks for all the good work!
